### PR TITLE
Add Spring Boot app with initial models for API metadata model

### PR DESCRIPTION
**Pull request by nahteb**
This is a proof of concept and not the actual model we're going to end up with, but any thoughts about the model are appreciated.

### CODE REVIEW

File: examples/java/spring-boot/build.gradle [hunk 1]
```diff
@@ -0,0 +1,8 @@
+plugins {
+  id 'uk.gov.api.spring-boot-conventions'
+}
+
+dependencies {
+  testImplementation 'org.springframework.boot:spring-boot-starter-test'
+  implementation 'org.springframework.boot:spring-boot-starter-web'
```
Comment (jamietanna) [inline examples/java/spring-boot/build.gradle:7]: Can we swap these around please? It reads a bit easier to see `api` then `implementation`, then anything `test*`
```suggestion
  implementation 'org.springframework.boot:spring-boot-starter-web'
  testImplementation 'org.springframework.boot:spring-boot-starter-test'
```

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/models/metadata/DataTest.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    @Test
+    void organisationIsSerialized() throws IOException {
+      assertThat(jsonContent).hasJsonPathStringValue("organisation", "some organisation");
+    }
+
+    @Test
+    void documentationUrlIsSerialized() throws IOException {
+      assertThat(jsonContent).hasJsonPathStringValue("documentation-url", "some documentation url");
+    }
+  }
```
Comment (jamietanna) [inline examples/java/spring-boot/src/test/java/uk/gov/api/springboot/models/metadata/DataTest.java:68]: IntelliJ shows these are not required:

```suggestion
    @Test
    void nameIsSerialized() {
      assertThat(jsonContent).extractingJsonPathStringValue("name").isEqualTo("some name");
    }

    @Test
    void descriptionIsSerialized() {
      assertThat(jsonContent)
          .extractingJsonPathStringValue("description")
          .isEqualTo("some description");
    }

    @Test
    void urlIsSerialized() {
      assertThat(jsonContent).extractingJsonPathStringValue("url").isEqualTo("some url");
    }

    @Test
    void contactIsSerialized() {
      assertThat(jsonContent).hasJsonPathStringValue("contact", "some contact");
    }

    @Test
    void organisationIsSerialized() {
      assertThat(jsonContent).hasJsonPathStringValue("organisation", "some organisation");
    }

    @Test
    void documentationUrlIsSerialized() {
      assertThat(jsonContent).hasJsonPathStringValue("documentation-url", "some documentation url");
    }
  }
```

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/models/metadata/DataTest.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
+package uk.gov.api.springboot.models.metadata;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (jamietanna) [inline examples/java/spring-boot/src/test/java/uk/gov/api/springboot/models/metadata/DataTest.java:4]: Not in use:
```suggestion
```

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/ApplicationTest.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package uk.gov.api.springboot;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (jamietanna) [inline examples/java/spring-boot/src/test/java/uk/gov/api/springboot/ApplicationTest.java:4]: Not in use:
```suggestion
```

**Review by jamietanna — CHANGES_REQUESTED** [review]
Sorry, these are a few things I forgot to mention while pairing!

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/models/metadata/ApiMetadataTest.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    @Test
+    void dataIsSerialized() throws IOException {
+      Data data = new Data();
+      ApiMetadata apiMetadata = new ApiMetadata();
+      apiMetadata.setData(data);
+      jsonContent = jacksonTester.write(apiMetadata);
+      assertThat(jsonContent).hasJsonPathMapValue("data");
+    }
```
Comment (jamietanna) [inline examples/java/spring-boot/src/test/java/uk/gov/api/springboot/models/metadata/ApiMetadataTest.java:40]: I'd recommend using line breaks to make this a given-when-then, and make this a little clearer in terms of what each set of steps are doing:

```suggestion
    void apiVersionIsSerialized() throws IOException {
      ApiMetadata apiMetadata = new ApiMetadata();
      apiMetadata.setApiVersion("v1-v2");
      
      jsonContent = jacksonTester.write(apiMetadata);
      
      assertThat(jsonContent).extractingJsonPathStringValue("api-version").isEqualTo("v1-v2");
    }

    @Test
    void dataIsSerialized() throws IOException {
      Data data = new Data();
      ApiMetadata apiMetadata = new ApiMetadata();
      apiMetadata.setData(data);
      
      jsonContent = jacksonTester.write(apiMetadata);
      
      assertThat(jsonContent).hasJsonPathMapValue("data");
    }
```

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/models/metadata/DataTest.java (hunk 3 above)
Comment (jamietanna) [reply examples/java/spring-boot/src/test/java/uk/gov/api/springboot/models/metadata/DataTest.java:4]:
```suggestion
import static org.junit.jupiter.api.Assertions.fail; 
```

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/models/metadata/DataTest.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.json.JsonTest;
+import org.springframework.boot.test.json.JacksonTester;
+import org.springframework.boot.test.json.JsonContent;
+
+@JsonTest
+class DataTest {
+
+  @Autowired private JacksonTester<Data> jacksonTester;
```
Comment (jamietanna) [inline examples/java/spring-boot/src/test/java/uk/gov/api/springboot/models/metadata/DataTest.java:19]:
```suggestion
  @Autowired
  private JacksonTester<Data> jacksonTester;
```

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/models/metadata/ApiMetadataTest.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.json.JsonTest;
+import org.springframework.boot.test.json.JacksonTester;
+import org.springframework.boot.test.json.JsonContent;
+
+@JsonTest
+public class ApiMetadataTest {
+
+  @Autowired private JacksonTester<ApiMetadata> jacksonTester;
```
Comment (jamietanna) [inline examples/java/spring-boot/src/test/java/uk/gov/api/springboot/models/metadata/ApiMetadataTest.java:18]:
```suggestion
  @Autowired 
  private JacksonTester<ApiMetadata> jacksonTester;
```

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/models/metadata/DataTest.java (hunk 6 above)
Comment (nahteb) [reply examples/java/spring-boot/src/test/java/uk/gov/api/springboot/models/metadata/DataTest.java:19]: this is something SpotlessCheck does, do we want to (/can we) override it?

Comment (jamietanna) [reply examples/java/spring-boot/src/test/java/uk/gov/api/springboot/models/metadata/DataTest.java:19]: Ah sorry, if this is Spotless doing it it's all good - my preference is that it's split, but the whole point of using Spotless with Google Code Style is to make sure my personal preferences go out the window :grin:

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/models/metadata/ApiMetadataTest.java (hunk 7 above)
Comment (jamietanna) [reply examples/java/spring-boot/src/test/java/uk/gov/api/springboot/models/metadata/ApiMetadataTest.java:18]: Per other comment, nope, as this is done by Spotless

**Review by jamietanna — DISMISSED** [review]
Looks good - would be good to get Charles/Arnau's thoughts before merge

**Review by jamietanna — APPROVED** [review]

**Pull request merged**